File: Dihya/ai/ai.py
"""
Module principal d’IA pour Dihya : NLP, ML, génération de code, fallback open source, intégration GPT/LLM, multilingue, sécurité, audit, tests.
"""

import os
from typing import Optional, Dict, Any

# IA providers (importez vos wrappers ou SDKs ici)
try:
    from openai import OpenAI
except ImportError:
    OpenAI = None

try:
    from llama_cpp import Llama
except ImportError:
    Llama = None

# Placeholders pour Mixtral/Mistral (à remplacer par vos wrappers réels)
class Mixtral:
    def generate(self, prompt, lang="fr"):
        return f"[Mixtral:{lang}] {prompt} (réponse simulée)"

class Mistral:
    def generate(self, prompt, lang="fr"):
        return f"[Mistral:{lang}] {prompt} (réponse simulée)"

# i18n & sécurité (à adapter selon vos modules)
def detect_language(text: str, supported=None) -> str:
    # Détection simple, à remplacer par un vrai algo
    if supported is None:
        supported = ["fr", "en", "ar", "tzr"]
    for lang in supported:
        if f"[{lang}]" in text:
            return lang
    return "fr"

def translate(text: str, target_lang: str) -> str:
    # Stub de traduction, à remplacer par un vrai service
    return f"[{target_lang}] {text}"

def sanitize_input(text: str) -> str:
    # Nettoyage basique, à renforcer selon vos besoins
    return text.replace("<script>", "").replace("</script>", "")

def sanitize_output(text: str) -> str:
    # Nettoyage basique, à renforcer selon vos besoins
    return text.strip()

def get_user_role(user_id: str) -> str:
    # À remplacer par une vraie gestion des rôles
    return "admin"

def check_permission(role: str, action: str) -> bool:
    # À adapter selon votre politique de rôles
    allowed = {
        "admin": ["ai:generate"],
        "user": ["ai:generate"],
        "contributor": ["ai:generate"]
    }
    return action in allowed.get(role, [])

def log_ai_request(data: Dict[str, Any]):
    # Logging/audit (à brancher sur votre SIEM si besoin)
    print(f"[AI-REQ] {data}")

def log_ai_response(data: Dict[str, Any]):
    print(f"[AI-RESP] {data}")

SUPPORTED_LANGUAGES = ["fr", "en", "ar", "tzr"]

class DihyaAIConfig:
    OPENAI_ENABLED = bool(os.getenv("OPENAI_API_KEY"))
    OPENAI_API_KEY = os.getenv("OPENAI_API_KEY", "")
    OPENAI_MODEL = os.getenv("OPENAI_MODEL", "gpt-4o")
    FALLBACK_MIXTRAL = True
    FALLBACK_LLAMA = True
    FALLBACK_MISTRAL = True
    MAX_TOKENS = 2048
    TEMPERATURE = 0.7
    ALLOWED_ROLES = ["admin", "user", "contributor"]

def select_engine(preferred: Optional[str] = "openai") -> str:
    if preferred == "openai" and DihyaAIConfig.OPENAI_ENABLED and OpenAI:
        return "openai"
    if DihyaAIConfig.FALLBACK_MIXTRAL:
        return "mixtral"
    if DihyaAIConfig.FALLBACK_LLAMA:
        return "llama"
    if DihyaAIConfig.FALLBACK_MISTRAL:
        return "mistral"
    raise RuntimeError("Aucun moteur IA disponible")

def generate_ai_response(
    prompt: str,
    user_id: str,
    lang: Optional[str] = None,
    role: Optional[str] = None,
    engine: Optional[str] = None
) -> Dict[str, Any]:
    """
    Génère une réponse IA souveraine, multilingue, sécurisée.
    :param prompt: Texte utilisateur (multi-langue)
    :param user_id: ID utilisateur
    :param lang: Langue cible (auto si absent)
    :param role: Rôle utilisateur (admin, user, etc.)
    :param engine: Moteur IA préféré
    :return: dict {text, lang, engine}
    """
    safe_prompt = sanitize_input(prompt)
    user_role = role or get_user_role(user_id)
    if not check_permission(user_role, "ai:generate"):
        raise PermissionError("Permission refusée")
    detected_lang = lang or detect_language(safe_prompt, SUPPORTED_LANGUAGES)
    ai_engine = select_engine(engine)

    log_ai_request({
        "user_id": user_id,
        "prompt": safe_prompt,
        "lang": detected_lang,
        "engine": ai_engine
    })

    ai_response = ""
    if ai_engine == "openai" and OpenAI:
        client = OpenAI(api_key=DihyaAIConfig.OPENAI_API_KEY)
        completion = client.chat.completions.create(
            model=DihyaAIConfig.OPENAI_MODEL,
            messages=[{"role": "user", "content": safe_prompt}],
            max_tokens=DihyaAIConfig.MAX_TOKENS,
            temperature=DihyaAIConfig.TEMPERATURE,
            user=user_id
        )
        ai_response = completion.choices[0].message.content
    elif ai_engine == "mixtral":
        ai_response = Mixtral().generate(safe_prompt, lang=detected_lang)
    elif ai_engine == "llama" and Llama:
        llama = Llama()
        ai_response = llama.generate(safe_prompt, lang=detected_lang)
    elif ai_engine == "mistral":
        ai_response = Mistral().generate(safe_prompt, lang=detected_lang)
    else:
        raise RuntimeError("Aucun moteur IA disponible")

    safe_output = sanitize_output(ai_response)
    log_ai_response({
        "user_id": user_id,
        "response": safe_output,
        "lang": detected_lang,
        "engine": ai_engine
    })

    # Traduction si besoin
    final_output = safe_output
    if detected_lang and not safe_output.startswith(f"[{detected_lang}]"):
        final_output = translate(safe_output, detected_lang)

    return {
        "text": final_output,
        "lang": detected_lang,
        "engine": ai_engine
    }

# Exemple d’utilisation (à commenter en prod)
if __name__ == "__main__":
    res = generate_ai_response(
        prompt="Génère un template métier pour la gestion RH.",
        user_id="demo-user",
        lang="fr",
        role="admin"
    )
    print(res)

# Exemples de fonctions à inclure :
# - analyse_text_multilingue()
# - generer_code_metier()
# - fallback_llm_open_source()
# - audit_ia_usage()
# - tests_unitaires_ia()

# ...

File: Dihya/ai/ai.js
/**
 * Dihya AI Core Module
 * Ultra avancé, multilingue, souverain, extensible, sécurisé.
 * Supporte fallback IA open source (Mixtral, LLaMA, Mistral), multi-stack, multi-rôle.
 * Prêt pour la production, la démo, la contribution.
 *
 * @module DihyaAI
 */

const { OpenAI } = require('openai');
const { LlamaCpp } = require('llama-cpp');
const { Mixtral } = require('mixtral-js');
const { Mistral } = require('mistral-js');
const { detectLanguage, translate } = require('./i18n');
const { getUserRole, checkPermission } = require('../auth/roles');
const { logAIRequest, logAIResponse } = require('../utils/audit');
const { sanitizeInput, sanitizeOutput } = require('../utils/security');

const SUPPORTED_LANGUAGES = ['fr', 'en', 'ar', 'tzr'];

/**
 * Configuration IA souveraine et fallback
 */
const config = {
  openai: {
    enabled: process.env.OPENAI_API_KEY ? true : false,
    apiKey: process.env.OPENAI_API_KEY || '',
    model: process.env.OPENAI_MODEL || 'gpt-4o',
  },
  fallback: {
    mixtral: true,
    llama: true,
    mistral: true,
  },
  maxTokens: 2048,
  temperature: 0.7,
  allowedRoles: ['admin', 'user', 'contributor'],
};

/**
 * Sélection dynamique du moteur IA selon la souveraineté et la disponibilité.
 */
async function selectEngine(preferred = 'openai') {
  if (preferred === 'openai' && config.openai.enabled) return 'openai';
  if (config.fallback.mixtral) return 'mixtral';
  if (config.fallback.llama) return 'llama';
  if (config.fallback.mistral) return 'mistral';
  throw new Error('Aucun moteur IA disponible');
}

/**
 * Génère une réponse IA souveraine, multilingue, sécurisée.
 * @param {Object} params
 * @param {string} params.prompt - Texte utilisateur (multi-langue)
 * @param {string} params.userId - ID utilisateur
 * @param {string} [params.lang] - Langue cible (auto si absent)
 * @param {string} [params.role] - Rôle utilisateur (admin, user, etc.)
 * @param {string} [params.engine] - Moteur IA préféré
 * @returns {Promise<{text: string, lang: string, engine: string}>}
 */
async function generateAIResponse({ prompt, userId, lang, role, engine }) {
  // Sécurité & audit
  const safePrompt = sanitizeInput(prompt);
  const userRole = await getUserRole(userId);
  if (!checkPermission(userRole, 'ai:generate')) {
    throw new Error('Permission refusée');
  }
  const detectedLang = lang || detectLanguage(safePrompt, SUPPORTED_LANGUAGES);
  const aiEngine = await selectEngine(engine);

  logAIRequest({ userId, prompt: safePrompt, lang: detectedLang, engine: aiEngine });

  let aiResponse = '';
  if (aiEngine === 'openai') {
    const openai = new OpenAI({ apiKey: config.openai.apiKey });
    const completion = await openai.chat.completions.create({
      model: config.openai.model,
      messages: [{ role: 'user', content: safePrompt }],
      max_tokens: config.maxTokens,
      temperature: config.temperature,
      user: userId,
    });
    aiResponse = completion.choices[0].message.content;
  } else if (aiEngine === 'mixtral') {
    const mixtral = new Mixtral();
    aiResponse = await mixtral.generate(safePrompt, { lang: detectedLang });
  } else if (aiEngine === 'llama') {
    const llama = new LlamaCpp();
    aiResponse = await llama.generate(safePrompt, { lang: detectedLang });
  } else if (aiEngine === 'mistral') {
    const mistral = new Mistral();
    aiResponse = await mistral.generate(safePrompt, { lang: detectedLang });
  } else {
    throw new Error('Aucun moteur IA disponible');
  }

  // Sécurité sortie & audit
  const safeOutput = sanitizeOutput(aiResponse);
  logAIResponse({ userId, response: safeOutput, lang: detectedLang, engine: aiEngine });

  // Traduction si besoin
  let finalOutput = safeOutput;
  if (detectedLang && !safeOutput.startsWith(`[${detectedLang}]`)) {
    finalOutput = await translate(safeOutput, detectedLang);
  }

  return {
    text: finalOutput,
    lang: detectedLang,
    engine: aiEngine,
  };
}

/**
 * Exemple d’utilisation (à supprimer en prod)
 */
// (async () => {
//   const res = await generateAIResponse({
//     prompt: 'Génère un template métier pour la gestion RH.',
//     userId: 'demo-user',
//     lang: 'fr',
//     role: 'admin',
//   });
//   console.log(res);
// })();

module.exports = {
  generateAIResponse,
  selectEngine,
  SUPPORTED_LANGUAGES,
  config,
};

// ai.js – Dihya AI Core (Node/JS)

/**
 * Module principal d’IA pour Dihya : NLP, ML, génération de code, fallback open source, intégration GPT/LLM, multilingue, sécurité, audit, tests.
 */

// ...existing code or à compléter selon la stack IA choisie...

// Exemples de fonctions à inclure :
// - analyseTextMultilingue()
// - genererCodeMetier()
// - fallbackLlmOpenSource()
// - auditIaUsage()
// - testsUnitairesIa()

// ...
